[Do not merge] Add code for Microsoft.Azure.Services.AppAuthentication library

varunsh-coder opened the pull request:
<!-- DO NOT DELETE THIS TEMPLATE -->

## Description
<!--
Please add an informative description that covers that changes made by the pull request.

If you are regenerating your SDK based off of a new swagger spec, please add the link to the corresponding swagger spec pull request that has been merged in the azure-rest-api-specs repository
-->

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [ ] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [ ] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [ ] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

Based on discussion with @shahabhijeet, creating a PR to test CI. Will wait for @shahabhijeet to review this in first week of October.

Commit:
Move code to GitHub

First commit of the Microsoft.Azure.Services.AppAuthentication library

Commit:
Update test cert env variable name

Commit:
Update test cert env variable name

Commit:
Add readme file

Commit:
Update test cases

Exclude integration test cases from running after build

msftclas commented:
@varunsh-msft,
Thanks for your contribution as a Microsoft full-time employee or intern. You do not need to sign a CLA.
_Thanks,
Microsoft Pull Request Bot_

AArnott commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.IntegrationTests/Microsoft.Azure.Services.AppAuthentication.IntegrationTests.csproj:
```diff
@@ -0,0 +1,69 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  
```
nit: why so many empty lines?

shahabhijeet commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.IntegrationTests/Microsoft.Azure.Services.AppAuthentication.IntegrationTests.csproj:
```diff
@@ -0,0 +1,69 @@
+    
+		<TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
+  
+		<SignAssembly>True</SignAssembly>
```
@varunsh-msft you  do not need this property, it's already been added as part of common properties references for the entire repo

shahabhijeet commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.IntegrationTests/Microsoft.Azure.Services.AppAuthentication.IntegrationTests.csproj:
```diff
@@ -0,0 +1,69 @@
+    	
+		<AssemblyOriginatorKeyFile>$(LibraryToolsFolder)\MSSharedLibKey.snk</AssemblyOriginatorKeyFile>
+    
+		<DelaySign>True</DelaySign>
```
redundant property, please remove

shahabhijeet commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.IntegrationTests/Microsoft.Azure.Services.AppAuthentication.IntegrationTests.csproj:
```diff
@@ -0,0 +1,69 @@
+
+  
+
+	<PropertyGroup Condition=" '$(TargetFramework)'=='net452' ">
```
the compiler constant as well as property group is part of common targets/props. Please remove

shahabhijeet commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.TestCommon/Microsoft.Azure.Services.AppAuthentication.TestCommon.csproj:
```diff
@@ -0,0 +1,61 @@
+    
+		<TargetFrameworks>netcoreapp1.1;net452</TargetFrameworks>
+  
+		<SignAssembly>True</SignAssembly>
```
@varunsh-msft if you model your project file after one of the existing project file, that will help eliminating redundant properties.

shahabhijeet commented on src/SdkCommon/AppAuthentication/Azure.Services.AppAuthentication.Unit.Tests/Microsoft.Azure.Services.AppAuthentication.Unit.Tests.csproj:
```diff
@@ -0,0 +1,65 @@
+    
+		<PackageReference Include="Newtonsoft.Json" Version="9.0.1" />
+    
+		<PackageReference Include="xunit" Version="2.2.0" />
```
@varunsh-msft please model your test project from one of the eixsitng test project and remove redundant properties

Commit:
Update httpclient usage in msi token provider

Unit test cases will send an httpclient with a mocked handler. Actual
library use will result in use of a default static httpclient.

yane3628 committed:
Add test that confirms an exception is thrown with the Azure CLI fails to get token

Commit:
Update cache test

Remove the need to wait in the test and ensure calls are made
concurrently.

yane3628 committed:
Merge branch 'AppAuthentication' of https://github.com/varunsh-msft/azure-sdk-for-net into AppAuthenticationUpdatedTests

yane3628 committed:
Add test case to check when azureAdInstance is null or empty the appropriate exception is thrown

yane3628 committed:
Add test that confirms an exception is thrown with the connection string has an invlalid developer tool

yane3628 committed:
Update test that checks if AzureAdInstance is empty

yane3628 committed:
Add test that confirms an exception is thrown with the connection string has an invalid value for RunAs

yane3628 committed:
Add test that confirms success if the connection string when creating a tokenprovider ends with an extra semi-colon and space, that the parser ignores the extra space and continues

yane3628 committed:
Add test that confirms an exception is thrown when the connection string is not valid

yane3628 committed:
Add test that confirms an exception is thrown when the connection string has a repeating key

yane3628 committed:
Add test that confirms an exception is thrown when the connection string does not contain '='

yane3628 committed:
Add test that confirms an exception is thrown when the connection string is null or empty

yane3628 committed:
Add test that confirms an exception is thrown when the connection string doesn't haave at least one required attribute

yane3628 committed:
Update test that ignores an extra semi-colon and white space at the end of the connection string

yane3628 committed:
Add test that confirms an exception is thrown when the connection string doesn't have a valid cert location

yane3628 committed:
Add test that confirms an exception is thrown when the AppId is null or empty when making a ClientCertificateAzureServiceTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the storeLocation is null or empty when making a ClientCertificateAzureServiceTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the cert subject name or thumbprint is null or empty when making a ClientCertificateAzureServiceTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the store location is invalid when making a ClientCertificateAzureServiceTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the clientId is null or empty when making a ClientSecretAccessTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the clientSecret is null or empty when making a ClientSecretAccessTokenProvider

yane3628 committed:
Add test that confirms an exception is thrown when the resource is null or empty when getting an access token

yane3628 committed:
Remove redundant check for resourceID null or empty in private method GetAccessTokenAsynImpl

yane3628 committed:
Add test that confirms an environment variable is used when the connectionstring is not specified when making an AzureServiceTokenProvider

Commit:
Update clientId for Integrated Windows Authentication

This is specific clientid for the library.

yane3628 committed:
Add test that confirms an exception is thrown when token was failed to be aquired by all of the specified providers

yane3628 committed:
Add test that confirms an exception is thrown when token was failed to be aquired by a cert

yane3628 committed:
Add test that confirms an exception is thrown when there is a parsing error for the token when getting token through MSI

yane3628 committed:
Add test to check ensure exception is thrown when access token is missing in MSI response json object. Changes to exception messages in get token function in MSI provider and parsers

yane3628 committed:
Change MsiFailed enum to throw http request exception instead of time out

yane3628 committed:
Add test to ensure exception is thrown when HttpResquestException occurs

yane3628 committed:
Add test to ensure exception is thrown when access token is not in the expected format when acquiring through cert using authentication context

yane3628 committed:
Add assert to Azure CLI failure test to check the exception message from the token provider

yane3628 committed:
Update AzureServiceTokenProviderFactory to throw an ArgumentException instead of FormatException and change test to assert for ArgumentException

yane3628 committed:
Change environment variable from string literal to value defined in Constants.cs

yane3628 committed:
Update comments to indicate expected failure

yane3628 committed:
Fix typo git status

yane3628 committed:
Fix typo for MockAuthenticationContext enum

yane3628 committed:
Fix order of cases in switch statement for Mock MSI, moved MsiAppServicesSuccess to fall through MsiAzureVmSucess

yane3628 committed:
Add assert in json parse error test to check that the deserialization error from the exception message is found

yane3628 committed:
Add assert in access token null test to check that the object reference not set to instance of an object is found in the exception

yane3628 committed:
Remove AzureCliConnectionString environment variable in the dispose function of AzureServiceTokenProviderTests

yane3628 committed:
Add assert to check the exception of invalid token causing expected failure of getting token from cert

yane3628 committed:
Add check for null accessToken in AccessToken.Parse and change assert messages for expected exception in MsiResponseMissingTokenTest

Commit:
Merge pull request #1 from varunsh-msft/AppAuthenticationUpdatedTests

App authentication updated tests

shahabhijeet commented:
@varunsh-msft as we talked earlier this week. We will revisit merging this code next month. Until then please address conflicts in this PR and reopen the PR.
Closing it for now.
